contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -900,8 +895,85 @@ clearly described and the permissions issues should be resolved.
+
+The operator is expected to know the version, group, and resource name of the object it needs to retrieve the value from, and can use the discovery client or construct the API path directly.
+
+### References which can point to multiple resource types
+
+Multi-kind object references are used when there is a bounded set of valid resource types that a reference can point to.
+
+As with a single-kind object reference, the operator can supply missing fields, provided that the fields that are present are sufficient to uniquely identify the object resource type among the set of supported types.
+
+```yaml
+# guidance for the field name is the same as a single resource.
+fooRef:
+    group: sns.services.k8s.aws
+    resource: topics
+    name: foo
+    namespace: foo-namespace
+```
+
+Although not always necessary to help a controller identify a resource type, “group” is included to avoid ambiguity when the resource exists in multiple groups. It also provides clarity to end users and enables copy-pasting of a reference without the referenced type changing due to a different controller handling the reference.
+
+#### Controller behavior
+
+The operator can store a map of (group,resource) to the version of that resource it desires. From there, it can construct the full path to the resource, and retrieve the object.
+
+### Generic object reference
+
+A generic object reference is used when the desire is to provide a pointer to some object to simplify discovery for the user. For example, this could be used to reference a target object for a `core.v1.Event` that occurred.
+
+With a generic object reference, it is not possible to extract any information about the referenced object aside from what is standard (e.g. ObjectMeta). Since any standard fields exist in any version of a resource, it is possible to not include version in this case:
+
+```yaml
+fooObjectRef:
+    group: operator.openshift.io
+    resource: OpenShiftAPIServer
+    name: cluster
+    # namespace is unset if the resource is cluster-scoped, or lives in the 
+    # same namespace as the referrer.
+```
+
+#### Controller behavior
+
+The operator would be expected to find the resource via the discovery client (as the version is not supplied).
+
+### Field reference
+
+A field reference is used when the desire is to extract a value from a specific field in a referenced object.
+
+Field references differ from other reference types, as the operator has no knowledge of the object prior to the reference. Since the schema of an object can differ for different versions of a resource, this means that a “version” is required for this type of reference.
+
+```yaml
+fooFieldRef:
+   version: v1 # version of the resource
+   # group is elided in the ConfigMap example, since it has a blank group in the OpenAPI spec.
+   resource: configmaps
+   fieldPath: data/foo
+```
+
+The fieldPath should point to a single value, and use JSON Pointer syntax to specify the desired field, from the root of the object.
```

Comment:
one note about this line: there already is a [convention around field selection syntax](https://github.com/kubernetes/community/blob/e8f0fd99f5793e31c6ae624be9c3677fc3aecc38/contributors/devel/sig-architecture/api-conventions.md#selecting-fields), that seems to recommend using javascript field access syntax rather than JSON Pointer.

I think this probably also needs to hashed out, but I'm thinking it might be better to hash that out in a follow-up PR given the size of this one.

Reply:
Oops, I pasted my comment then saw this.  +100.

Reply:
Perhaps reference the spec? https://datatracker.ietf.org/doc/html/rfc6901

Reply:
@lavalamp I don't disagree with the use of JSONPointer, but since there is already an existing section around using the Javascript field selection syntax, how do you feel about using that syntax for this PR, and I'll follow up with another one to propose changing it (fieldPath and the section recommending Javascript selection syntax) to JSONPointer?

Just don't want to stall this PR on this design point. Otherwise I'll bring this to a sig or some slack channel to make headway.

Reply:
I could also see us adopting this spec but we'd want to go back and update the fieldpath logic to allow either syntax.  But only if:

* We think this syntax is materially better for some treason
* We can unambiguously tell them apart (or the overlapping parts are identical in meaning)

We also *just* added a label to namespaces called "metadata.name", which I think we'd want to add an equivalent of the new syntax.

IOW - I don't mind adopting new syntax AS LONG AS we can explain why it's worth the (considerable) effort to do so.

Reply:
Personally, I feel like I have the "JSON Path is not formalized so I don't want it in the API" conversation often enough that I'd really prefer to get JSONPointer in the first draft.

Reply:
Are we going to go fix all the places that allow the old syntax?  Who is signing up for that work?  Alternatively, we can just document our own notation since we know that we don't have to be as general as JSONPointer

Reply:
> Alternatively, we can just document our own notation since we know that we don't have to be as general as JSONPointer

We might end-up having to do that, especially if we want to refer to something that is in an associative lists (can be done with json path but it's fairly awkward).

Reply:
> want to refer to something that is in an associative list

Ugh, we probably do. Maybe that means using the format in structured-merge-diff.

Reply:
@toumorokoshi Do you have a lot of field references in your CRD? I'm curious if you can give some more examples.

Reply:
Our format is "optimized" for sets of paths, rather than declaring a single path, so I'm not sure.

Reply:
Do we have any places where we use the home-grown fieldpath syntax to descend into an associative list?  I see metadata.name, .namespace, .labels, .labels['string'], etc but none of those are list-maps.

Are there other places that parse fieldpath-style strings without using the fieldpath lib?

Reply:
@apelisse we have a format for individual paths, though! There are some examples in this test file: https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/fieldpath/path_test.go

Reply:
I think we only have the printing implemented at the moment, so it's only half done for this use case. However it should handle all the weird sub-cases of our api, such as referencing a single element by-value in a list which is really a set. :)

Reply:
using structured-merge-diff does seem like a great way to give some breathing room. I don't think it's entirely clear what all the types of schemas that fieldpath will have to navigate are, and a syntax that is owned by a K8S SIG means we can extend the syntax as needed.

> Are we going to go fix all the places that allow the old syntax? Who is signing up for that work? Alternatively, we can just document our own notation since we know that we don't have to be as general as JSONPointer

I can take a follow-up item to find uses of the old syntax if we decide to move forward.

Reply:
Can I get a recap?

Are we changing our bespoke fieldpath syntax to something?  If so, what syntax and what is the rationale?  If not, do we need to extend the bespoke syntax to cover some new use case?

Reply:
The current proposal is to try to change the [recommendation](https://github.com/kubernetes/community/blob/bffe980b66c9bd743dc04759ad3dd72285fcab60/contributors/devel/sig-architecture/api-conventions.md#selecting-fields) from "JavaScript syntax for accessing that field" into something a bit more formalized, like JSON Pointer. The main driver of that opinion is @lavalamp. I'll let @lavalamp elaborate if I'm incorrect, but IIUC it's because that recommendation is fairly loose and undefined. 

@apelisse is in this conversation too, but I can't tell if you have a specific opinion on this.

I [wrote a quick doc to summarize my understanding of the options](https://docs.google.com/document/d/1-qHEZ8vgvYf-ykOnSJtG8jOFor_-SZr2t1SEwGX4jT8/edit#) around field selectors.

Based on that research, my suggestion is:

1. keep the current guidance (not directly stating, but implying, to use the [k8s/pkg/fieldpath](https://github.com/kubernetes/kubernetes/blob/master/pkg/fieldpath/fieldpath.go#L46) code), and pick this discussion up in a separate PR / KEP.
2. If 1 is not an option, I would still argue we keep using the fieldpath code referenced above, but expand it for more use cases.

My thought on 2 is that:

- fieldpath code is already embedded and is being used for implementations of `FieldPath`. migrating these use cases would take time, in addition to breaking users of some of these (e.g. `EnvVarSource.valueFrom.FieldRef.FieldPath`)
- we can define a spec to handle the "undefined or loose" part of the problem with fieldpath.
- owning our own propietary format enables us to extend it to handle things that javascript-style field selectors may not be good at, like associate lists (where identity is determined by a field in the element, rather than it's position in a list). Since it's not too far away from the other bespoke format (merge diff), we may even be able to unify it eventually.

Reply:
I'm not expressing a strong opinion *yet*, because I think we should do that:
> 1. keep the current guidance (not directly stating, but implying, to use the k8s/pkg/fieldpath code), and pick this discussion up in a separate PR / KEP.

Reply:
updated the PR to just point to the current guidance for now. If there's any guidance discussed in a SIG meeting I'll update here.

Reply:
@lavalamp ping to you that I've switched guidance to recommend the existing section for now, with a commitment to you to follow up on normalizing our fieldPath options a bit (see [this document for some notes](https://docs.google.com/document/d/1-qHEZ8vgvYf-ykOnSJtG8jOFor_-SZr2t1SEwGX4jT8/edit#)). 

Does that sound ok?